File: go-sdk/kestra_api_client/promotion_targets_api.go
package kestra_api_client

import "context"

// PromotionTargetsAPI covers /api/v1/{tenant}/promotion-targets. The controller
// is gated by the PROMOTION_TARGET resource and the FEATURE_PROMOTE license
// feature, so routes answer 403 when the feature is disabled.
type PromotionTargetsAPI struct {
	baseAPI
}

// ListPromotionTargets returns a page of promotion targets. Backs GET
// /api/v1/{tenant}/promotion-targets.
func (a *PromotionTargetsAPI) ListPromotionTargets(ctx context.Context, tenant string, page, size *int, sort []string, filters []SearchFilter) (map[string]interface{}, error) {
	params := buildQueryParams("page", page, "size", size)
	appendRepeatedParam(params, "sort", sort)
	appendFilterParams(params, filters)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "promotion-targets"), nil, params)
}

// PromotionTarget returns a single promotion target by id. Backs GET
// /api/v1/{tenant}/promotion-targets/{id}.
func (a *PromotionTargetsAPI) PromotionTarget(ctx context.Context, id, tenant string) (map[string]interface{}, error) {
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "promotion-targets", id), nil, nil)
}

// CreatePromotionTarget creates a promotion target. `body` is an
// ApiCreatePromotionTargetRequest. Backs POST /api/v1/{tenant}/promotion-targets.
func (a *PromotionTargetsAPI) CreatePromotionTarget(ctx context.Context, tenant string, body interface{}) (map[string]interface{}, error) {
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "POST", tenantPath(tenant, "promotion-targets"), body, nil)
}

// UpdatePromotionTarget updates a promotion target. Backs PUT
// /api/v1/{tenant}/promotion-targets/{id}.
func (a *PromotionTargetsAPI) UpdatePromotionTarget(ctx context.Context, id, tenant string, body interface{}) (map[string]interface{}, error) {
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "PUT", tenantPath(tenant, "promotion-targets", id), body, nil)
}

// DeletePromotionTarget deletes a promotion target. Backs DELETE
// /api/v1/{tenant}/promotion-targets/{id}.
func (a *PromotionTargetsAPI) DeletePromotionTarget(ctx context.Context, id, tenant string) error {
	return a.doVoidJSON(ctx, "DELETE", tenantPath(tenant, "promotion-targets", id), nil, nil)
}

// TestPromotionTarget tests an unsaved promotion-target definition. The optional
// `id` is sent as a query param (not a path segment) to reuse a stored token.
// Backs POST /api/v1/{tenant}/promotion-targets/test.
func (a *PromotionTargetsAPI) TestPromotionTarget(ctx context.Context, tenant string, body interface{}, id *string) (*ApiTestConnectionResponse, error) {
	params := buildQueryParams("id", id)
	return doJSON[*ApiTestConnectionResponse](&a.baseAPI, ctx, "POST", tenantPath(tenant, "promotion-targets", "test"), body, params)
}

// TestSavedPromotionTarget tests a stored promotion target's connection. Backs
// POST /api/v1/{tenant}/promotion-targets/{id}/test.
func (a *PromotionTargetsAPI) TestSavedPromotionTarget(ctx context.Context, id, tenant string) (*ApiTestConnectionResponse, error) {
	return doJSON[*ApiTestConnectionResponse](&a.baseAPI, ctx, "POST", tenantPath(tenant, "promotion-targets", id, "test"), nil, nil)
}

// PromotionTargetFlowSource fetches the source of a flow as it exists on the
// promotion target. Backs GET /api/v1/{tenant}/promotion-targets/{id}/flow-source.
func (a *PromotionTargetsAPI) PromotionTargetFlowSource(ctx context.Context, id, tenant, namespace, flowId string, revision *int) (map[string]interface{}, error) {
	params := buildQueryParams("namespace", namespace, "flowId", flowId, "revision", revision)
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "GET", tenantPath(tenant, "promotion-targets", id, "flow-source"), nil, params)
}

// PromotionTargetFlowHashes returns per-flow content hashes on the promotion
// target for a batch of flow references. `body` is an ApiFlowHashBatch.Request
// ({flows: [{namespace, flowId}]}). Backs POST
// /api/v1/{tenant}/promotion-targets/{id}/flow-hashes.
func (a *PromotionTargetsAPI) PromotionTargetFlowHashes(ctx context.Context, id, tenant string, body interface{}) (map[string]interface{}, error) {
	return doJSON[map[string]interface{}](&a.baseAPI, ctx, "POST", tenantPath(tenant, "promotion-targets", id, "flow-hashes"), body, nil)
}

File: python/python-sdk/kestrapy/api/promotion_targets_api.py
from typing import Any, Dict, List, Optional

from kestrapy.base_api import BaseApi
from kestrapy.models.api_create_promotion_target_request import ApiCreatePromotionTargetRequest
from kestrapy.models.api_flow_hash_batch_request import ApiFlowHashBatchRequest
from kestrapy.models.api_flow_hashes_response import ApiFlowHashesResponse
from kestrapy.models.api_promotion_target_response import ApiPromotionTargetResponse
from kestrapy.models.api_test_connection_response import ApiTestConnectionResponse
from kestrapy.models.api_update_promotion_target_request import ApiUpdatePromotionTargetRequest
from kestrapy.models.paged_results_api_promotion_target_response import (
    PagedResultsApiPromotionTargetResponse,
)
from kestrapy.models.query_filter import QueryFilter


class PromotionTargetsApi(BaseApi):
    """Promotion targets (`/api/v1/{tenant}/promotion-targets`).

    A promotion target is a remote Kestra instance/tenant that flows can be
    promoted to. The controller is gated by the ``PROMOTION_TARGET`` resource
    and the ``FEATURE_PROMOTE`` licence feature, so every route answers 403 when
    the feature is disabled.
    """

    def list_promotion_targets(
        self,
        tenant: str,
        page: Optional[int] = None,
        size: Optional[int] = None,
        sort: Optional[List[str]] = None,
        filters: Optional[List[QueryFilter]] = None,
    ) -> PagedResultsApiPromotionTargetResponse:
        path = self._tenant_path(tenant, "promotion-targets")
        params = list(self._build_query_params(page=page, size=size).items())
        self._append_repeated_param(params, "sort", sort)
        self._append_filter_params(params, filters)
        return self._json_request("GET", path, PagedResultsApiPromotionTargetResponse, params=params)

    def promotion_target(self, id: str, tenant: str) -> ApiPromotionTargetResponse:
        path = self._tenant_path(tenant, "promotion-targets", id)
        return self._json_request("GET", path, ApiPromotionTargetResponse)

    def create_promotion_target(
        self, tenant: str, body: ApiCreatePromotionTargetRequest
    ) -> ApiPromotionTargetResponse:
        path = self._tenant_path(tenant, "promotion-targets")
        return self._json_request("POST", path, ApiPromotionTargetResponse, body=body)

    def update_promotion_target(
        self, id: str, tenant: str, body: ApiUpdatePromotionTargetRequest
    ) -> ApiPromotionTargetResponse:
        path = self._tenant_path(tenant, "promotion-targets", id)
        return self._json_request("PUT", path, ApiPromotionTargetResponse, body=body)

    def delete_promotion_target(self, id: str, tenant: str) -> None:
        path = self._tenant_path(tenant, "promotion-targets", id)
        self._void_request("DELETE", path)

    def test_promotion_target(
        self,
        tenant: str,
        body: ApiCreatePromotionTargetRequest,
        id: Optional[str] = None,
    ) -> ApiTestConnectionResponse:
        """Test an unsaved promotion-target definition. ``id`` is sent as a query
        param (not a path segment) to reuse a stored token."""
        path = self._tenant_path(tenant, "promotion-targets", "test")
        params = self._build_query_params(id=id)
        return self._json_request("POST", path, ApiTestConnectionResponse, body=body, params=params)

    def test_saved_promotion_target(self, id: str, tenant: str) -> ApiTestConnectionResponse:
        path = self._tenant_path(tenant, "promotion-targets", id, "test")
        return self._json_request("POST", path, ApiTestConnectionResponse)

    def promotion_target_flow_source(
        self,
        id: str,
        tenant: str,
        namespace: str,
        flow_id: str,
        revision: Optional[int] = None,
    ) -> Dict[str, Any]:
        path = self._tenant_path(tenant, "promotion-targets", id, "flow-source")
        params = self._build_query_params(namespace=namespace, flowId=flow_id, revision=revision)
        return self._raw_json_request("GET", path, params=params)

    def promotion_target_flow_hashes(
        self, id: str, tenant: str, body: ApiFlowHashBatchRequest
    ) -> ApiFlowHashesResponse:
        path = self._tenant_path(tenant, "promotion-targets", id, "flow-hashes")
        return self._json_request("POST", path, ApiFlowHashesResponse, body=body)
